### PR TITLE
142686  - Validating config file

### PR DESCRIPTION
**Pull request by chaital14503**
PR consists of following changes for [PBI142686](https://dev.azure.com/ukhydro/Exchange%20Set%20Service/_workitems/edit/142686)

- Code changes to validate config file and UTs

Kindly review the same.

### CODE REVIEW

**Review by shirinbt14926 — APPROVED** [review]
LGTM

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Validation/ConfigValidator.cs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        ValidationResult Validate(BessConfig configurationSetting);
+    }
+
+    public class ConfigValidator : AbstractValidator<BessConfig>, IConfigValidator
+    {
+        public ConfigValidator()
+        {
+            RuleFor(config => config.Name)
+               .Must(name => !string.IsNullOrEmpty(name?.Trim()))
+               .WithMessage("Attribute is missing or value not provided")
```
Comment (harshal11869) [inline UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Validation/ConfigValidator.cs:21]: As this validation message text is same for different validation can we make constant string and use that across ?
```suggestion
               .WithMessage("Attribute is missing or value is not provided")
```

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs [hunk 2]
```diff
@@ -1,8 +1,10 @@
 ﻿using System.Diagnostics.CodeAnalysis;
+using FluentValidation.Results;
 using Microsoft.Extensions.Logging;
 using NCrontab;
 using Newtonsoft.Json;
 using Newtonsoft.Json.Linq;
+using UKHO.BESS.ConfigurationService.Validation;
```
Comment (harshal11869) [inline UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs:7]: Can we improve Stryker score of this class ?

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Validation/ConfigValidatorTests.cs [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+                Tags = null,
+                ReadMeSearchFilter = "",
+                BatchExpiryInDays = 0,
+                IsEnabled = "yes"
+            };
+
+            TestValidationResult<BessConfig> result = configValidator.TestValidate(bessConfig);
+
+            result.ShouldHaveValidationErrorFor(x => x.Name)
+                .WithErrorMessage("Attribute is missing or value not provided");
```
Comment (harshal11869) [inline UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Validation/ConfigValidatorTests.cs:126]: As this validation message text is same for different validation can we make constant string and use that across ?
```suggestion
                .WithErrorMessage("Attribute is missing or value is not provided");
```

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs [hunk 4]
```diff
@@ -16,80 +18,134 @@ public class ConfigurationService : IConfigurationService
...
 
-                        bool isValidJson = IsValidJson(content, fileName);
+                        ValidationResult results = configValidator.Validate(deserializedConfig);
 
-                        if (isValidJson)
+                        if (!results.IsValid)
                         {
-                            IList<BessConfig> bessConfig = JsonConvert.DeserializeObject<List<BessConfig>>(content)!;
+                            configsWithInvalidAttributeCount++;
+                            string errors = string.Empty;
```
Comment (harshal11869) [inline UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs:67]: Can we use StringBuilder here ?
```suggestion
                            var errors = new StringBuilder();
```

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Validation/ConfigValidator.cs (hunk 1 above)
Comment (chaital14503) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Validation/ConfigValidator.cs]: Resolved. Thanks

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Validation/ConfigValidatorTests.cs (hunk 3 above)
Comment (chaital14503) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Validation/ConfigValidatorTests.cs:126]: Resolved. Thanks

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs (hunk 4 above)
Comment (chaital14503) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs]: Resolved. Thanks

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs (hunk 2 above)
Comment (chaital14503) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs:8]: Resolved. Thanks

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Validation/ConfigValidatorTests.cs [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+                Tags = null,
+                ReadMeSearchFilter = "",
+                BatchExpiryInDays = 0,
+                IsEnabled = "yes"
+            };
+
+            TestValidationResult<BessConfig> result = configValidator.TestValidate(bessConfig);
+
+            result.ShouldHaveValidationErrorFor(x => x.Name)
+                .WithErrorMessage("Attribute is missing or value is not provided");
```
Comment (harshal11869) [inline UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Validation/ConfigValidatorTests.cs:126]: As below validation messages text are same for different validations can we make constant string and use that across ?
1. "Attribute is missing or value is not provided"
2. "AllowedUsers and AllowedUserGroups both attribute values are not provided. Either of them should be provided"

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Validation/ConfigValidatorTests.cs (hunk 5 above)
Comment (chaital14503) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Validation/ConfigValidatorTests.cs]: resolved. thanks

**Review by harshal11869 — APPROVED** [review]
Looks Good.

**Review by Mayuresh-Gawde — APPROVED** [review]
LGTM from QA!!

**Review by DanOwenUKHO — APPROVED** [review]
LGTM

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Validation/ConfigValidator.cs [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+            {
+                if (!c.AllowedUsers.Any() || c.AllowedUsers.All(user => user?.Trim() is null or "") || c.AllowedUsers.Contains(string.Empty) || c.AllowedUsers.Contains(null))
+                    return false;
+            }
+
+            if (c.AllowedUserGroups != null && c.AllowedUsers != null)
+            {
+                if (c.AllowedUsers.All(user => user?.Trim() is null or "") && c.AllowedUserGroups.All(group => group?.Trim() is null or ""))
+                    return false;
+            }
```
Comment (rsj-github) [inline UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Validation/ConfigValidator.cs:121]: so this too could be written out neater in an if else

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs [hunk 7]
```diff
@@ -16,80 +19,140 @@ public class ConfigurationService : IConfigurationService
...
         public void ProcessConfigs()
         {
             try
             {
-                IList<BessConfig> bessConfigs = new List<BessConfig>();
+                var configsInContainer = azureBlobStorageClient.GetConfigsInContainer();
+                if (!configsInContainer.Any())
+                {
+                    logger.LogWarning(EventIds.BessConfigsNotFound.ToEventId(), "Bess configs not found | _X-Correlation-ID : {CorrelationId}", CommonHelper.CorrelationID);
+                }
```
[the remark this answers is not in the document]
Comment (Nadeem14456) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs]: Resolved.

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Validation/ConfigValidator.cs [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+            {
+                result.Errors.Add(new ValidationFailure("IsEnabled", "Bess config for file - " + context.InstanceToValidate.FileName + ", will be skipped for exchange set creation since the attribute value is set to “no”."));
+                return false;
+            }
+            else if (context.InstanceToValidate.IsEnabled?.ToLower() != "yes")
+            {
+                result.Errors.Add(new ValidationFailure("IsEnabled", "Attribute is missing or value is not provided. Expected value is either Yes or No."));
+                return false;
+            }
+            return true;
```
[the remark this answers is not in the document]
Comment (Nadeem14456) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Validation/ConfigValidator.cs:170]: We can follow the above suggestion; I believe that using guard clauses or early return from a function terminates both the if statement and the entire function.
@DanOwenUKHO and @harshal11869, what are your recommendations?

**Review by rsj-github — CHANGES_REQUESTED** [review]
from the unit test side did not see where the validation on the cell confirmation

**Comment by Nadeem14456** [comment]
> from the unit test side did not see where the validation on the cell confirmation

This PBI does not include validation of enc cells. It's covered in Sprint 4 PBI => https://dev.azure.com/ukhydro/Exchange%20Set%20Service/_workitems/edit/142688

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Validation/ConfigValidator.cs (hunk 6 above)
Comment (Nadeem14456) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Validation/ConfigValidator.cs:121]: We had tried this, but due to validation requirements for (AllowedUsers and AllowedUserGroups) we had to move this code to nested if.

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Validation/ConfigValidator.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (harshal11869) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Validation/ConfigValidator.cs:170]: Both condition is different so i guess we need to check separately.

**Review by rsj-github — APPROVED** [review]

**Pull request merged**